### PR TITLE
docs: follow guidelines for repo standards

### PR DESCRIPTION
**Pull request by cpate4**
https://broadworkbench.atlassian.net/browse/IA-4869

this is an implementation of:
https://broadworkbench.atlassian.net/wiki/spaces/TLR/pages/2952298505/GitHub+Repo+Standards#Repository-automation-standards

### CODE REVIEW

File: .github/CODEOWNERS
Comment (pshapiro4broad) [inline .github/CODEOWNERS:1]: This file supports `#` comments so maybe a good initial content would be a header comment and an example line. Actually maybe `platform-foundation-codeowners` should be the code owners here?
```
# Lines starting with '#' are comments.
# Each line is a file pattern followed by one or more owners.

# These owners will be the default owners for everything in the repo.
*       @DataBiosphere/platform-foundation-codeowners 

# Order is important. The last matching pattern has the most precedence.
# So if a pull request only touches javascript files, only these owners
# will be requested to review.
#*.js    @octocat @github/js

# You can also use email addresses if you prefer.
#docs/*  docs@example.com
```

File: CONTRIBUTING.md
Comment (pshapiro4broad) [inline CONTRIBUTING.md:1]: Maybe a starting point is to add notes here about how to contribute to the template repo itself?

File: bin/build
Comment (pshapiro4broad) [inline bin/build:1]: This project can be built, run and has tests. So I wouldn't expect these files to be empty even in the template repo.

**Review by pshapiro4broad — COMMENTED** [review]
I think we can both provide better examples and replace empty/placeholder text with text that documents this project itself.

Also if this should have a ticket on it maybe it should be a PF one?

**Comment by cpate4** [comment]
Thanks @pshapiro4broad - I converted the PR to a draft. Still a work in progress (filling out all of the stuff from our standard guidelines relevant to this repo)

File: .github/CODEOWNERS
Comment (cpate4) [reply .github/CODEOWNERS:1]: ... yup. it is a placeholder so I could go back to it. 
thank you for the recommendation re: `platform-foundation-codeowners`.

File: bin/build
Comment (cpate4) [reply bin/build:1]: they won't be empty when I am done ... don't worry 😉

File: CONTRIBUTING.md
Comment (cpate4) [reply CONTRIBUTING.md:1]: document details (from our standards guide) forthcoming! - 🚀

File: .gitignore
Comment (cpate4) [inline .gitignore:17]: before I delete this ... 
I am curious to understand why `bin/` is currently bing ignored?

... and happy to move the `bin/` directory to `scripts/` (or something else).
thank you in advance! 🤓

Comment (pshapiro4broad) [reply .gitignore:17]: No idea about `bin/`, maybe some tools use it as an output directory?

Regarding the setup of a `scripts` directory, I think that makes sense, and I'd advocate to set it up as a gradle subproject so you can run `shellcheck` and `black` on it on PRs.
https://github.com/DataBiosphere/terra-data-catalog/blob/main/scripts/build.gradle
https://github.com/DataBiosphere/terra-data-catalog/blob/main/.github/workflows/build-and-test.yml#L20

Comment (cpate4) [reply .gitignore:17]: we can talk more about this tomorrow at our meeting. 
i'd like to find/create a standard place where people can find these scripts without having to look in different places.

File: .github/CODEOWNERS
Comment (cahrens) [inline .github/CODEOWNERS:11]: This isn't about permissions. Isn't it just saying that the CODEOWNERS file requires review by platform-foundation-codeowners? And isn't that already covered by the default statement at line 9?

Comment (pshapiro4broad) [reply .github/CODEOWNERS:11]: AFAIK the presence of CODEOWNERS does not by itself enforce that they're required for a review. That requires setting the branch protections in the repo config. It also doesn't mean that they're tagged on reviews, this also must be enabled separately. My understanding is that this file is a configuration setting that _can_ be used but isn't' by default.

Also this line does nothing since this path is already included in `*` which is set above, so I would remove it.

File: service/publishing.gradle
Comment (pshapiro4broad) [inline service/publishing.gradle:31]: I think we should go the other direction, and also remove `filesModificationTime` from the config, so that jib builds are reproducible. https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#why-is-my-image-created-48-years-ago https://reproducible-builds.org

File: scripts/setup
Comment (pshapiro4broad) [inline scripts/setup:16]: Would we also want to verify the version of tools, e.g. `java -version | ...`?

File: service/publishing.gradle
Comment (cpate4) [reply service/publishing.gradle:31]: happy to remove it ... I read that link, too, and found it weirdly bizarre all of our docker images from this build process were considered "built at epoch" (read 'decades ago')

File: .github/CODEOWNERS
Comment (cahrens) [reply .github/CODEOWNERS:11]: That's true, the number of reviews and way they are selected is done through branch protections in the repo config.

It is possible to override assignments because the last specified does win out, but since they are the same, I see no point in explicitly assigning CODEOWNERS to the group.

File: service/publishing.gradle
Comment (cpate4) [reply service/publishing.gradle:31]: ... as for `filesModificationTime` it is specifically called out that it is there to prevent ui caching. maybe we should be wiring in ... something else? githsha?

File: .github/CODEOWNERS
Comment (cpate4) [reply .github/CODEOWNERS:11]: All good! I can remove it. I was explicitly [calling it out](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-and-branch-protection) as a reflection of this section in the docs (see specifying CODEOWNERS specifically).

Comment (cahrens) [reply .github/CODEOWNERS:11]: If your repo is set up to require reviews by CODEOWNERS, it's true that having the explicit assignment to the CODEOWNERS file is the most secure because it guarantees that all edits to the file are reviewed by the group. But I don't think we really need that-- this isn't so much about locking down the repo as it as about helping ensure folks know who to contact. And if you have multiple teams contributing to the repo (like terra-ui), you expect CODEOWNERS to be edited by multiple teams. https://github.com/DataBiosphere/terra-ui/blob/dev/.github/CODEOWNERS

File: scripts/setup
Comment (cpate4) [reply scripts/setup:16]: that is the "beauty" of it ... we can definitely add that if we are interested in doing so. we can be as strict or lenient as we need to be

File: CONTRIBUTING.md
Comment (cahrens) [inline CONTRIBUTING.md:20]: The directory is "scripts", not "bin". See also line 22.

File: scripts/setup
Comment (cahrens) [inline scripts/setup:1]: Don't we want scripts to have the .sh extension?

Comment (cpate4) [reply scripts/setup:1]: It wasn't a part of our original proposal, but happy to modify it

File: CONTRIBUTING.md
Comment (cpate4) [reply CONTRIBUTING.md:26]: [mermaid.js](https://mermaid.js.org)
... and GitHub made it available through [their markdown](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/). 

It allows us to add and iterate on [diagrams like this](https://github.com/DataBiosphere/terra-java-project-template/blob/IA-XXXX_document-standards/CONTRIBUTING.md#developer-convenience-scripts)

Comment (cahrens) [reply CONTRIBUTING.md:26]: Oh, cool! I didn't think to look at the rendered version.

File: scripts/setup
Comment (cpate4) [reply scripts/setup:1]: ... wanted to tag @mflinn-broad on this for the discussion purpose of it. I know he is working on various devops repos on this, too. for context, I made the `.sh` adds locally and didn't want to surprise him in case he didn't have suffixes on his scripts.

Comment (mflinn-broad) [reply scripts/setup:1]: Mine don't have the .sh extension currently. As a general convention on devops we don't include .sh. But it doesn't really matter to me one way or the other. If folks want .sh I'm fine with that. I think it matters more that we just pick one and stick to it more so than which we actually pick.

Comment (cpate4) [reply scripts/setup:1]: my original thought here was ...
it doesn't matter if it's python or bash or ??? - the command was the same and its only requirement is that it should be executable. 

adding the extensions means we will be following `setup.sh` and `setup.py` and `setup.rs` and ... look good?

Comment (mflinn-broad) [reply scripts/setup:1]: That was also my original thought

Comment (cpate4) [reply scripts/setup:1]: @cahrens - It felt unusual to me at first, too - and totally happy to discuss more if you feel strongly about it. You have a valid point I am definitely not attached to it. 

And to support your point and concern - let's say we say, "write these entry points this way (`setup` / `build` / `run`)." - there is nothing to preventing anyone from encountering the situation where people feel strongly about adding the file suffixes onto it. My then counterpoint to this would be to support (the potential of) having more templatize-ation (not sure if that's a word, but going to roll with it) around our build processes. That would be "the selling point" - if you will. Ultimately, we can afford anything to be parameterized, but then I hit myself with "convention over configuration"-conversation. :sigh: - and yes, my brain is constantly like this.  😬 

Did we follow all of that? - it's totally easy to add back on or take off - it'll be harder after we release this into the wild - so thank you for entertaining the conversation with me - 😇

File: scripts/init.sh
Comment (cahrens) [inline scripts/init.sh:1]: @cpate4 so `init` and `utils` we are going to enforce as bash scripts (have extension), but the others not?

Comment (cahrens) [reply scripts/init.sh:1]: Or is the idea that these 2 files are things that we are not expecting folks to need their own implementation of?

Do you desire people to add to their own version of `init.sh`, or should anything custom live in `setup`?

Comment (cpate4) [reply scripts/init.sh:1]: ... and to your last point, it is the that these 2 things we are not expecting folks to need their own implementations of. they are things the repo provided by way of help / organization - but the entry points are the standards. 

that being said, you are 100% correct. and using my own observations here, if it:
- isn't intuitive, and
- it isn't clear for the next person coming along, then
- it probably isn't a very good pattern

Comment (cpate4) [reply scripts/init.sh:1]: @mflinn-broad / @pshapiro4broad - because you both weighted in before ... @cahrens is rightly pointing out these subtle nuances. thoughts? opinions? how do we make this clearer so people aren't sorting through a directory structure to figure out their entry points?

Comment (cpate4) [reply scripts/init.sh:1]: ... I am reviewing the PR one last time this morning, and in doing so, and in support of @cahrens's comments (again), I originally made `run-db` follow the `run` syntax, but with these additional scripts, should we?

in our world of black-and-white, we ran right into the gray 🤪

Comment (mflinn-broad) [reply scripts/init.sh:1]: My thinking with scripts like this is that no extension means it's an executable, .sh means it's a library that can be `source`'d into an executable. I realize that may not be apparent to everyone though. There isn't really a standard way to distinguish that type of thing with shell scripts that I'm aware of.

My opinion is that we should only be prescriptive about the `setup`, `build` and `run`. If teams want to add additional supporting logic, I think they should be encouraged to do so however works best for them, as long as core setup, build, and run interface is preserved.

I originally suggested the inclusion of utils.sh to @cpate4 because it contains a couple functions that are going to be generally useful in pretty much any script. ie a standardized way to get formatted log out put and to check that all the tools required to execute the script are installed on the user's machine.

Comment (mflinn-broad) [reply scripts/init.sh:1]: I don't have a strong opinion on the inclusion of an init.sh. I didn't include it in my implementation broadinstitute/yale because there just aren't really any environment variables to set there.

**Review by mflinn-broad — APPROVED** [review]
Thanks for doing this! I don't really have any additional comments beyond what has already been said. I think this gives us a great starting point for getting this effort shared with the wider org. 

We can always come back and make updates to this as we learn more from trying this out in the wild.

File: scripts/run
Comment (cahrens) [inline scripts/run:13]: Can we go with "integration"? I think it will be hard to remember "integ" given that it's not a standard abbreviation.

File: scripts/build
Comment (cahrens) [inline scripts/build:21]: We don't need skip_tests anymore in build because it's now part of `run`.

File: scripts/run
Comment (cpate4) [reply scripts/run:13]: if these were options, I would adopt the `--t` and `--i` in a heartbeat, but this is a positional argument for `RUNTYPE`. moved to `integration`

File: scripts/build
Comment (cpate4) [reply scripts/build:21]: this is needed as a part of the `build_docker()` and `build_project()` functions to skip tests while building

File: scripts/utils.sh
Comment (cahrens) [inline scripts/utils.sh:2]: I think we should add a note about contributing changes back to this repo (https://github.com/DataBiosphere/terra-landing-zone-service/pull/421#discussion_r1524015977).

File: scripts/build
Comment (cahrens) [inline scripts/build:15]: I still vote for "lint" being included.

Comment (cpate4) [reply scripts/build:15]: ... I'll look again, but I didn't see a "lint" step explicitly called out in the build anywhere. Would you happen to know where one might exist?

Comment (cahrens) [reply scripts/build:15]: I wasn't looking in this repo, just thinking of other repos (https://github.com/DataBiosphere/terra-landing-zone-service/pull/421/files#diff-1641870daa39645825488c81f1ba274a062d49f96635e19297b80f05a21582ad).

Comment (cpate4) [reply scripts/build:15]: ... would you recommend it being added as a no-op?

or can/should we add it when this repo has some sort of static linting / formatting / something added?

Comment (pshapiro4broad) [reply scripts/build:15]: Hi, sorry I haven't been around for the code reviews, I do have an opinion about this though.

I think the spotless check should be run as part of `build`. I don't think it's helpful to have a separate `lint` step. Although maybe that's where the `sonar` and `srcclr` steps could go?

File: scripts/init.sh
Comment (pshapiro4broad) [reply scripts/init.sh:1]: I don't spend much time working with bash or python but I would lean toward always adding an extension, even for `run`. If we have a `scripts` gradle subproject, we can configure it to run `shellcheck` and `black` on PRs, which I think would require that they have correct extensions. (unless these tools look at all files for `#!` preambles?)

This how I set up the `scripts` subproject in `catalog`: https://github.com/DataBiosphere/terra-data-catalog/blob/main/scripts/build.gradle

Comment (cpate4) [reply scripts/init.sh:1]: @pshapiro4broad - thanks for the feedback. [based on the referenced configs](https://github.com/DataBiosphere/terra-data-catalog/blob/main/scripts/build.gradle#L11) it is possible to configure gradle subprojects to include our scripts

File: scripts/build
Comment (cpate4) [reply scripts/build:15]: this is configured here for this project:
`buildSrc/src/main/groovy/bio.terra.java-common-conventions.gradle`

**Pull request merged**
